src/check_pack.c

```diff
@@ -33,13 +34,33 @@
 #include "check_impl.h"
 #include "check_pack.h"
 
+#ifdef HAVE_WINDOWS_H
+#include <windows.h>
+#endif
+
+#ifdef HAVE_SYNCHAPI_H
+#include <synchapi.h>
+#endif
+
+#if defined(HAVE_INIT_ONCE_BEGIN_INITIALIZE) && defined(HAVE_INIT_ONCE_COMPLETE)
+#define HAVE_WIN32_INIT_ONCE 1
+#endif
+
 #ifndef HAVE_PTHREAD
-#define pthread_mutex_lock(arg)
-#define pthread_mutex_unlock(arg)
 #define pthread_cleanup_push(f, a) {
 #define pthread_cleanup_pop(e) }
 #endif
 
+#ifndef HAVE_PTHREAD
+#if defined(HAVE_WIN32_INIT_ONCE)
+#define pthread_mutex_lock(arg) EnterCriticalSection(arg)
```

Comment:
Replacing the definition of the pthread methods seems a little magical to me. Reading the code later on it will not be obvious that a critical section is being used instead of pthreads.

I think it would be better to have a level of abstraction here about mutual exclusion. In Check's source there are several re-implementations of some APIs that may be missing or may use one of multiple implementations:

https://github.com/libcheck/check/tree/master/lib

The high resolution timer support is one example:

https://github.com/libcheck/check/blob/master/lib/timer_settime.c

There, if the high resolution calls are not available replacements are used.

Perhaps there can be some functions created that abstract away the details for pthread and critical sections.

Reply:
Makes sense, I can implement a basic pthread_mutex inside lib on top of Win32.